lock/src/main/java/org/infinispan/lock/impl/lock/RequestExpirationScheduler.java

```diff
@@ -0,0 +1,128 @@
+ * @since 9.2
+ */
+public class RequestExpirationScheduler {
+
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass(), Log.class);
+   private static final boolean trace = log.isTraceEnabled();
+
+   private final ScheduledExecutorService scheduledExecutorService;
+   private final ConcurrentMap<String, ScheduledRequest> scheduledRequests = new ConcurrentHashMap<>();
+
+   public RequestExpirationScheduler(ScheduledExecutorService scheduledExecutorService) {
+      this.scheduledExecutorService = scheduledExecutorService;
+   }
+
+   class ScheduledRequest {
+      private CompletableFuture<Boolean> request;
+      private ScheduledFuture<?> scheduledFuture;
+
+      public ScheduledRequest(CompletableFuture<Boolean> request, ScheduledFuture<?> scheduledFuture) {
+         this.request = request;
+         this.scheduledFuture = scheduledFuture;
+      }
+
+      public CompletableFuture<Boolean> getRequest() {
+         return request;
+      }
+
+      public ScheduledFuture<?> getScheduledFuture() {
+         return scheduledFuture;
+      }
+   }
+
+   /**
+    * Schedules a request for completion
+    *
+    * @param requestId, the unique identifier if the request
+    * @param request,   the request
+    * @param time,      time expressed in long
+    * @param unit,      {@link TimeUnit}
+    */
+   public void scheduleForCompletion(String requestId, CompletableFuture<Boolean> request, long time, TimeUnit unit) {
+      if (request.isDone()) {
+         if (trace) {
+            log.tracef("Request[%s] is not scheduled because is already done", requestId);
+         }
+         return;
+      }
+
+      if (scheduledRequests.containsKey(requestId)) {
+         if (trace) {
+            log.tracef("Request[%s] is not scheduled because it is already scheduled", requestId);
+         }
+         return;
+      }
+
+      runScheduling(requestId, request, time, unit);
+   }
+
+   //Visible for testing
+   Runnable runScheduling(String requestId, CompletableFuture<Boolean> request, long time, TimeUnit unit) {
+      if (trace) {
+         log.tracef("Request[%s] being scheduled to be completed in [%d, %s]", requestId, time, unit);
+      }
+
+      Runnable lambda = () -> {
+         request.complete(false);
+         scheduledRequests.remove(requestId);
+      };
+
+      ScheduledFuture<?> scheduledFuture = scheduledExecutorService.schedule(lambda, time, unit);
+
+      scheduledRequests.putIfAbsent(requestId, new ScheduledRequest(request, scheduledFuture));
+
+      return lambda;
+   }
+
+   /**
+    * Aborts the scheduled request
+    *
+    * @param requestId, unique identifier of the request
+    */
+   public void abortScheduling(String requestId) {
+      if (trace) {
+         log.tracef("Request[%s] abort scheduling", requestId);
+      }
+      ScheduledRequest scheduledRequest = scheduledRequests.get(requestId);
+      if (scheduledRequest != null && scheduledRequest.request.isDone()) {
```

Comment:
I am not sure if you saw my last comment about this, but I am fine if you don't remove it (more of a best practices kinda thing)

Reply:
ok